src/actions/editorialWorkflow.js

```diff
@@ -291,3 +326,66 @@ export function publishUnpublishedEntry(collection, slug) {
     });
   };
 }
+
+export const registerUnpublishedEntryDependency = unpublishedEntryRegisterDependency;
+
+export function getUnpublishedEntryDependencies (collection, slug) {
+  return (dispatch, getState) => {
+    dispatch(unpublishedEntryDependenciesRequest(collection, slug));
+
+    const getDepsPath = dep => [
+      "entities",
+      dep,
+      "metaData",
+      "dependencies",
+    ];
+
+    const getEventualDependencies = (paths, loadedDeps, state) =>
+       paths.filter(path => !loadedDeps.includes(path)).map((path) => {
+         const [pathCollectionName, pathSlug] = path.split(".");
+         const pathCollection = state.collections.get(pathCollectionName);
+         // Wait for the entry to load
+         return dispatch(loadUnpublishedEntry(pathCollection, pathSlug))
+         // Return the path at the end so we can use it in .thens later
+           .then(() => path);
+       });
+
+    const traverse = (collectedDeps, path) => {
+      const state = getState();
+
+      if (collectedDeps.includes(path)) {
+        return new Promise(resolve => resolve(collectedDeps));
+      }
+
+      const newDeps = collectedDeps.push(path);
+      const depsPath = getDepsPath(path);
+
+      if (!state.editorialWorkflow.hasIn(depsPath) ||
+          state.editorialWorkflow.getIn(depsPath).size === 0) {
+        // This path has no dependencies
+        return new Promise(resolve => resolve(newDeps));
+      }
+
+      const theseDependencies = state.editorialWorkflow.getIn(depsPath);
+
+      const eventualDeps = getEventualDependencies(theseDependencies, collectedDeps, state);
+
+      // This is a manual version of Bluebird's Promise.reduce
+      return eventualDeps.reduce((depsPromise, depPromise) =>
+        Promise.all([depsPromise, depPromise]).then(
+          ([deps, dep]) => traverse(deps, dep)),
+        new Promise(resolve => resolve(newDeps)));
+    };
+
+    // Begin traversal
+    return traverse(new Immutable.List(), `${ collection }.${ slug }`)
```

Comment:
If you use a Map rather than a List, you'll avoid the N lookup per iteration in the `traverse` call.